File: JavaScript/DailyCodeChallenge/260313_calculateParkingFee.js
//HAD TO USE GEMINI ASSISTANCE HERE AS WAS CAUGHT UP IN THE LOGIC OF USING A DATE TYPE


function calculateParkingFee(parkTime, pickupTime) {

  const [parkH, parkM] = parkTime.split(':').map(Number);
  const [pickH, pickM] = pickupTime.split(':').map(Number);
  
  let startMinutes = parkH * 60 + parkM;
  let endMinutes = pickH * 60 + pickM;

  let isOvernight = false;

  // 2. Handle the Midnight Cross
  if (endMinutes < startMinutes) {
    endMinutes += 24 * 60; // Add 24 hours in minutes
    isOvernight = true;
  }

  // 3. Calculate total hours (rounding up)
  const durationMinutes = endMinutes - startMinutes;
  const parkHours = Math.ceil(durationMinutes / 60);

  // 4. Calculate Costs
  // Rule: $3 per hour, but minimum of $5
  let baseCost = parkHours * 3;
  if (baseCost < 5) baseCost = 5;

  // Rule: Add $10 if it spanned overnight
  const totalCost = isOvernight ? baseCost + 10 : baseCost;

  return `$${totalCost}`;
}

// Test Case 5: 18:15 to 01:30
// Duration: 7 hours 15 mins -> 8 hours. 
// (8 * 3) + 10 = $34.
console.log(calculateParkingFee("18:15", "01:30")); // "$34"

// Test Case 6: 11:11 to 11:10 (Next day)
// Duration: 23 hours 59 mins -> 24 hours.
// (24 * 3) + 10 = $82.
console.log(calculateParkingFee("11:11", "11:10")); // "$82"

File: Python/Daily_Code_Challenge/260313-calculate-parking-fee.py
import math

def calculate_parking_fee(park_time, pickup_time):
    # 1. Split and convert to integers
    park_h, park_m = map(int, park_time.split(':'))
    pick_h, pick_m = map(int, pickup_time.split(':'))
    
    start_minutes = park_h * 60 + park_m
    end_minutes = pick_h * 60 + pick_m

    is_overnight = False

    # 2. Handle the Midnight Cross
    if end_minutes < start_minutes:
        end_minutes += 24 * 60  # Add 24 hours in minutes
        is_overnight = True

    # 3. Calculate total hours (rounding up)
    duration_minutes = end_minutes - start_minutes
    # math.ceil returns an integer in Python 3
    park_hours = math.ceil(duration_minutes / 60)

    # 4. Calculate Costs
    # Rule: $3 per hour, but minimum of $5
    base_cost = park_hours * 3
    if base_cost < 5:
        base_cost = 5

    # Rule: Add $10 if it spanned overnight
    total_cost = base_cost + 10 if is_overnight else base_cost

    return f"${total_cost}"

# --- Test Cases ---
# Test Case 5: 18:15 to 01:30 (8 hours @ $3 + $10 overnight)
print(calculate_parking_fee("18:15", "01:30"))  # Output: $34

# Test Case 6: 11:11 to 11:10 (24 hours @ $3 + $10 overnight)
print(calculate_parking_fee("11:11", "11:10"))  # Output: $82